src/plasma/plasma_client.cc

```diff
@@ -465,13 +469,67 @@ void plasma_contains(PlasmaConnection *conn, ObjectID obj_id, int *has_object) {
   }
 }
 
+void compute_block_hash(const unsigned char *data, int64_t nbytes, uint64_t *hash) {
+  XXH64_state_t hash_state;
+  XXH64_reset(&hash_state, XXH64_DEFAULT_SEED);
+  XXH64_update(&hash_state, data, nbytes);
+  *hash = XXH64_digest(&hash_state);
+}
+
+bool compute_object_hash_parallel(XXH64_state_t *hash_state,
+                                  const unsigned char *data,
+                                  int64_t nbytes) {
+#ifndef NUMTHREADS
+#define NUMTHREADS 8
+#endif
+  const uint64_t numthreads = NUMTHREADS;
+  uint64_t threadhash[numthreads+2];
+  //CHECK(numthreads > 0);
+  const uint64_t blocksz = 64; // cache block alignment (alternative: page size)
+  // Calculate the first and last aligned positions in the data stream.
+  unsigned char *databp = (unsigned char *)(((uint64_t)data + blocksz-1) & ~(blocksz-1));
+  unsigned char *dataep = (unsigned char *)(((uint64_t)data + nbytes) & ~(blocksz-1));
+  /* Calculate how many cache blocks we have to hash and divide them equally. */
+  const uint64_t numblocks = (((uint64_t)dataep - (uint64_t)databp)) / blocksz;
+  // uint64_t remainder = numblocks % numthreads;
+  // Update the end pointer
+  dataep = dataep - (numblocks % numthreads)*blocksz;
+  const uint64_t prefix = (uint64_t)databp - (uint64_t)data; // Bytes
+  const uint64_t suffix = (uint64_t)(data+nbytes) - (uint64_t)dataep; // Bytes
+  // Now data == | prefix | k*numthreads*blocksz | suffix |
+  // chunksz = k*blocksz => data == | prefix | numthreads*chunksz | suffix |
+  // Each thread gets a "chunk" of k blocks, except prefix and suffix threads.
+  uint64_t chunksz = ((uint64_t)dataep - (uint64_t)databp) / numthreads;// Bytes
+
+  std::vector<std::thread> threads;
+  // Start the prefix thread.
+  threads.push_back(std::thread(
+      compute_block_hash, data, prefix, &threadhash[0]));
+  for (int i = 1; i <= numthreads; i++) {
```

Comment:
@pcmoritz , I could do that, but in cases where we get aligned, well-behaved input, we won't even have to start the prefix and suffix threads, making the expected number of actual threads \in [numthreads; numthreads+2] in the general case. Alternatively, I could issue the prefix and suffix memcopy in the main thread, without spawning a thread for it. Would that be better? I felt that, having a guaranteed *tight* margin on the expected number of threads is sufficient.